server/WAPI.rb

```diff
@@ -0,0 +1,189 @@
+  # through this instance.  It will contain host and may anything else
+  # that will appear in front of every request.  E.g. It might contain
+  # https://woodpigeon.dsc.umich.edu:8243/StudentDashboard/v1 or just
+  # https://woodpigeon.dsc.umich.edu:8243 depending on how you choose
+  # to use it.
+
+  def initialize(application)
+    if application.nil?
+      msg = "No ESB Application values provided to WAPI initialize"
+      logger.warn msg
+      raise StandardError, msg
+    end
+
+    @token_server = application['token_server']
+    @api_prefix = application['api_prefix']
+    @key = application['key']
+    @secret = application['secret']
+    @token = application['token']
+
+    # ## special uniqname may be supplied for testing
+    @uniqname = application['uniqname']
+
+    @renewal = WAPI.build_renewal(@key, @secret)
+    logger.info("WAPI: #{__LINE__}: initialize WAPI with #{@api_prefix}")
+  end
+
+
+
+  ### Consider making this a separate class with helpful methods
+  ### to access portions of the result and to convert types.
+  def self.wrap_result(status, msg, result)
+    Hash["Meta" => Hash["httpStatus" => e,
+                        "Message" => msg],
+         "Result" => result]
+  end
+
+
+  def self.build_renewal(key, secret)
+    b64 = base64_key_secret(key, secret)
+    "Basic #{b64}"
+  end
+
+  def self.base64_key_secret(key, secret)
+    ks = "#{key}:#{secret}"
+    Base64.strict_encode64(ks)
+  end
+
+  # use instance information to generate the proper url
+  def format_url(request)
+    "#{@api_prefix}#{request}"
+  end
+
+  def do_request(request)
+    url=format_url(request)
+    #logger.debug "WAPI: do_request: url: #{url}"
+    r = Stopwatch.new(url)
+    r.start
+    begin
+      response = RestClient.get url, {:Authorization => "Bearer #{@token}",
+                                      :accept => :json,
+                                      :verify_ssl => true}
+
+      ## try to parse as json or send back a wrapped error
+      j = JSON.parse(response)
+      #logger.debug "WAPI: #{__LINE__}: do_request: esb response "+j.inspect
+
+      ## convert response code to integer if comes as a string
+      begin
+        # convert the response code to an integer
+        if j.has_key?('responseCode')
+          j['responseCode'] = j['responseCode'].to_i
+        end
+      rescue => err
+        logger.info "WAPI: #{__LINE__}: do_request: conversion error "+j.inspect
+        ## because of the error reset back to the original response.
+        j = JSON.parse(response)
+      end
+      ## get response code from nested response or from the original response
+      rc = j['responseCode'] || response.code
+      j = JSON.generate(j)
+      wrapped_response = WAPIResultWrapper.new(rc, "COMPLETED", j)
+    rescue Exception => exp
+      logger.debug "WAPI: #{__LINE__}: do_request: exception: "+exp.inspect
+      wrapped_response = WAPIResultWrapper.new(666, "EXCEPTION", exp)
```

Comment:
Instead of hard code the status, e.g. 666, 417, etc, those can be declared as constants with meaningful names.